yjit/src/codegen.rs

```diff
@@ -7871,21 +7882,28 @@ fn gen_send_iseq(
                 Counter::guard_send_block_arg_type,
             );
 
-            // If this is a forwardable iseq, adjust the stack size accordingly
-            let callee_ep = if forwarding {
-                -1 + num_locals + VM_ENV_DATA_SIZE as i32
-            } else {
-                -argc + num_locals + VM_ENV_DATA_SIZE as i32 - 1
-            };
-            let callee_specval = callee_ep + VM_ENV_DATA_INDEX_SPECVAL;
             if callee_specval < 0 {
                 // Can't write to sp[-n] since that's where the arguments are
                 gen_counter_incr(jit, asm, Counter::send_iseq_clobbering_block_arg);
                 return None;
             }
             let proc = asm.stack_pop(1); // Pop first, as argc doesn't account for the block arg
-            let callee_specval = asm.ctx.sp_opnd(callee_specval);
-            asm.store(callee_specval, proc);
+            let callee_specval_opnd = asm.ctx.sp_opnd(callee_specval);
+            asm.store(callee_specval_opnd, proc);
+
+            // When iseq_has_rest is true, there will be C calls (like rb_ary_dup) that
+            // can trigger GC before the callee frame is set up.
+            // The proc is stored above the current stack top, so we need to:
+            // - Nil-fill slots between stack top and specval (so GC doesn't see garbage)
+            // - Later, save SP with a larger offset to include the specval location
+            if iseq_has_rest && callee_specval > 0 {
+                nil_fill(
```

Comment:
The number of nils this fills scales based on the number of locals in the callee, which is not ideal since there's an unconditional nil fill for locals down the line.

I think it'd be better to side exit in the specific scenario localize the performance impact.